README.md
[![codecov](https://codecov.io/gh/Nucleus-Inc/day_schedule_list/branch/main/graph/badge.svg?token=JA1PRA7Z3V)](https://codecov.io/gh/Nucleus-Inc/day_schedule_list)
# Day Schedule List

Present your daily appointments on a widget similar to Google Calendar app day view.

## Installation

First, add `day_schedule_list` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/).

### Example

``` dart
import 'package:day_schedule_list/day_schedule_list.dart';
import 'package:flutter/material.dart';

class MyAppointment extends IntervalRange {
  MyAppointment({
    required this.title,
    required TimeOfDay start,
    required TimeOfDay end,
  }) : super(
          start: start,
          end: end,
        );

  final String title;
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  double heightOne = 120;
  double heightTwo = 180;
  double heightThree = 200;

  GlobalKey keyOne = GlobalKey();
  GlobalKey keyTwo = GlobalKey();
  GlobalKey keyThree = GlobalKey();

  final List<MyAppointment> myAppointments = [
    MyAppointment(
      title: 'Appointment 1',
      start: const TimeOfDay(hour: 8, minute: 40),
      end: const TimeOfDay(hour: 9, minute: 40),
    ),
    MyAppointment(
      title: 'Appointment 2',
      start: const TimeOfDay(hour: 11, minute: 0),
      end: const TimeOfDay(hour: 11, minute: 59),
    ),
    MyAppointment(
      title: 'Appointment 3',
      start: const TimeOfDay(hour: 14, minute: 15),
      end: const TimeOfDay(hour: 15, minute: 0),
    ),
    MyAppointment(
      title: 'Appointment 4',
      start: const TimeOfDay(hour: 16, minute: 10),
      end: const TimeOfDay(hour: 17, minute: 20),
    ),
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: const Text('My Schedule'),
        ),
        body: DayScheduleListWidget<MyAppointment>(
          hourHeight: 100,
          minimumMinuteInterval: MinuteInterval.thirty,
          appointmentMinimumDuration: MinuteInterval.thirty,
          referenceDate: DateTime.now(),
          appointments: myAppointments,
          dragIndicatorColor: Colors.red,
          updateAppointDuration: _updateAppointmentDuration,
          createNewAppointmentAt:
              (IntervalRange? interval, DayScheduleListWidgetErrors? error) {
            if (error != null || interval == null) {
              ScaffoldMessenger.of(context).showSnackBar(const SnackBar(
                content: Text(
                  'Failed to create new appointment',
                ),
              ));
            } else {
              setState(() {
                myAppointments.add(
                  MyAppointment(
                      title: 'New appointment',
                      start: interval.start,
                      end: interval.end),
                );
              });
            }
          },
          appointmentBuilder: _buildItem,
          unavailableIntervals: [
            IntervalRange(
                start: const TimeOfDay(hour: 0, minute: 0),
                end: const TimeOfDay(hour: 8, minute: 30)),
            IntervalRange(
                start: const TimeOfDay(hour: 12, minute: 0),
                end: const TimeOfDay(hour: 13, minute: 15)),
            IntervalRange(
                start: const TimeOfDay(hour: 18, minute: 0),
                end: const TimeOfDay(hour: 22, minute: 30))
          ],
        ));
  }

  Widget _buildItem(
      BuildContext context, MyAppointment appointment, double height) {
    Color color = height > 30 ? Colors.white : Colors.grey;
    return Card(
      margin: const EdgeInsets.symmetric(
        vertical: 1,
      ),
      color: Theme.of(context).primaryColor,
      child: Padding(
        padding: const EdgeInsets.all(5.0),
        child: Text(
          appointment.title,
          style: Theme.of(context).textTheme.caption?.copyWith(
                color: color,
              ),
        ),
      ),
    );
  }

  Future<bool> _updateAppointmentDuration(
      MyAppointment appointment, IntervalRange newInterval) {
    setState(() {
      appointment.start = newInterval.start;
      appointment.end = newInterval.end;
    });

    ///Save on server or locally the change and inform the success or not
    return Future.value(true);
  }
}

```

CHANGELOG.md
# 0.0.1
### Features
* List appointments.
* Gesture to change schedule duration.
* Gesture to change schedule initial time of day.
* Definition of unavailable time intervals.

# 0.0.2
* Fix some warnings

# 0.0.3
* Fix some warnings
* Improve imports

# 0.1.0
* Update README.md
* Package documentation
* Unit tests
* Add parameters to customize drag indicator widget
* Issue: Add ScrollController parameter
* Issue: Add Feature to change start date of appointment without changing its end date
* Add github action for test and upload code coverage to CodeCov service

# 0.2.0
* Issue: Change minimum time interval of appointment from 30 minutes to 15 minutes
* Issue: Add tap gesture to enable and disable edition mode
* Issue: Make possible to build child widget of Appointment based on duration of appointment
* Issue: Make possible to tap on some available time interval to create a new appointment

# 0.3.0
* Issue: Expand touch area of widget to change duration of appointment
* Issue: Change time interval to update appointment duration from 1 min to 15 min

# 0.4.0
* Issue: Dart code metrics
* Issue: 1 hour interval should contain 4 appointments of 15 min, but it is not happening

# 0.4.1
* Bug fix

# 0.4.2
* Issue: Fix error when defining first and last available hours
* Issue: Create possibility to customize drag indicator widget
* Issue: Add Read only mode

# 0.4.3
* Issue: Scroll day schedule list when changing schedule start time
* Issue: Improve internal classes organization

# 0.4.4
* Issue: Correção do bug na scroll na lista dos agendamentos, para funcionar em diferentes tamanhos de telas.

# 0.4.5
* Issue: Make it possible to add a new widget next to the main one

# 0.4.6
* Issue: Make it possible to change color of hours

example/lib/home_page.dart
import 'package:day_schedule_list/day_schedule_list.dart';
import 'package:flutter/material.dart';

class MyAppointment extends IntervalRange {
  MyAppointment({
    required this.title,
    required TimeOfDay start,
    required TimeOfDay end,
  }) : super(
          start: start,
          end: end,
        );

  final String title;
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key}) : super(key: key);
  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  double heightOne = 120;
  double heightTwo = 180;
  double heightThree = 200;

  GlobalKey keyOne = GlobalKey();
  GlobalKey keyTwo = GlobalKey();
  GlobalKey keyThree = GlobalKey();

  final List<MyAppointment> myAppointments = [
    MyAppointment(
      title: 'Appointment 1',
      start: const TimeOfDay(hour: 9, minute: 0),
      end: const TimeOfDay(hour: 9, minute: 15),
    ),
    MyAppointment(
      title: 'Appointment 2',
      start: const TimeOfDay(hour: 9, minute: 15),
      end: const TimeOfDay(hour: 9, minute: 30),
    ),
    MyAppointment(
      title: 'Appointment 3',
      start: const TimeOfDay(hour: 9, minute: 30),
      end: const TimeOfDay(hour: 9, minute: 45),
    ),
    MyAppointment(
      title: 'Appointment 4',
      start: const TimeOfDay(hour: 9, minute: 45),
      end: const TimeOfDay(hour: 10, minute: 0),
    ),
    MyAppointment(
      title: 'Appointment 5',
      start: const TimeOfDay(hour: 11, minute: 0),
      end: const TimeOfDay(hour: 11, minute: 59),
    ),
    // MyAppointment(
    //   title: 'Appointment 3',
    //   start: const TimeOfDay(hour: 14, minute: 15),
    //   end: const TimeOfDay(hour: 15, minute: 0),
    // ),
    // MyAppointment(
    //   title: 'Appointment 4',
    //   start: const TimeOfDay(hour: 16, minute: 10),
    //   end: const TimeOfDay(hour: 17, minute: 20),
    // ),
  ];

  @override
  void initState() {
    super.initState();
  }

  void listener() {
    debugPrint('Changed');
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: const Text('My Schedule'),
        ),
        body: DayScheduleListWidget<MyAppointment>(
          hourHeight: 100,
          minimumMinuteInterval: MinuteInterval.fifteen,
          appointmentMinimumDuration: MinuteInterval.fifteen,
          referenceDate: DateTime.now(),
          appointments: myAppointments,
          dragIndicatorColor: Colors.white,
          timeOfDayColor: Colors.brown,
          updateAppointDuration: _updateAppointmentDuration,
          optionalChildWidthLine: 30,
          optionalChildLine: (context, appointment, height) => Container(
            width: 30,
            height: height - 1,
            decoration: BoxDecoration(
              color: Colors.yellow,
              borderRadius: BorderRadius.circular(5),
            ),
            child: IconButton(
              onPressed: () {},
              icon: const Icon(Icons.abc_outlined),
            ),
          ),
          createNewAppointmentAt:
              (IntervalRange? interval, DayScheduleListWidgetErrors? error) {
            if (error != null || interval == null) {
              ScaffoldMessenger.of(context).showSnackBar(const SnackBar(
                content: Text(
                  'Failed to create new appointment',
                ),
              ));
            } else {
              setState(() {
                myAppointments.add(
                  MyAppointment(
                      title: 'New appointment',
                      start: interval.start,
                      end: interval.end),
                );
              });
            }
          },
          appointmentBuilder: _buildItem,
          unavailableIntervals: [
            IntervalRange(
                start: const TimeOfDay(hour: 0, minute: 0),
                end: const TimeOfDay(hour: 8, minute: 30)),
            IntervalRange(
                start: const TimeOfDay(hour: 12, minute: 0),
                end: const TimeOfDay(hour: 13, minute: 15)),
            IntervalRange(
                start: const TimeOfDay(hour: 18, minute: 30),
                end: const TimeOfDay(hour: 22, minute: 30))
          ],
        ));
  }

  Widget _buildItem(
      BuildContext context, MyAppointment appointment, double height) {
    Color color = height > 30 ? Colors.white : Colors.grey;
    return Card(
      margin: const EdgeInsets.symmetric(
        vertical: 1,
      ),
      color: Theme.of(context).primaryColor,
      child: Padding(
        padding: const EdgeInsets.all(5.0),
        child: Text(
          appointment.title,
          style: Theme.of(context).textTheme.caption?.copyWith(
                color: color,
              ),
        ),
      ),
    );
  }

  Future<bool> _updateAppointmentDuration(
      MyAppointment appointment, IntervalRange newInterval) {
    setState(() {
      appointment.start = newInterval.start;
      appointment.end = newInterval.end;
    });

    ///Save on server or locally the change and inform the success or not
    return Future.value(true);
  }
}
